# Web Scraping from OLX

## Scraping Mercedes Benz Car Price and Specification

In [1]:
# Library for scraping process

import numpy as np
import pandas as pd
from urllib.request import urlopen
from bs4 import BeautifulSoup

In [2]:
# scrapping script

Title = []
Price = []
Linked = []
Merek = []
Varian = []
Model = []
Tahun = []
Jarak_tempuh = []
Tipe_bbm = []
Warna =[]
Transmisi = []
Kapasitas_mesin = []
fixed_parameters = ['Merek','Varian', 'Model', 'Tahun', 'Jarak tempuh', 'Tipe bahan bakar', 'Warna', 'Transmisi', 'Kapasitas mesin']

pages = 1
while pages < 100:
    try :
        html = urlopen("https://www.olx.co.id/mobil-bekas_c198/q-mercedes-benz?page=%s" %(pages)).read()
        soup = BeautifulSoup(html, "lxml")
        produk = soup.find_all("li","EIR5N")
        for p in produk:
            if (p.find("span","_89yzn") != None) :
                title_temp = p.find("span","_2tW1I").get_text()
                price_temp = p.find("span","_89yzn").get_text()
                link = "https://www.olx.co.id"+p.find("a",href=True)['href']

                html_item = urlopen(link).read()
                soup_item = BeautifulSoup(html_item, "lxml")
                parameters_item = soup_item.find_all("span","_25oXN")
                produk_item = soup_item.find_all("span","_2vNpt")

                parameters = []  
                for param in parameters_item:
                    parameters.append(param.get_text())

                list_item = []
                for item in produk_item:
                    list_item.append(item.get_text())


                new_parameters = [item for item in parameters  if item in fixed_parameters ]

                if len(new_parameters) == len(fixed_parameters):
                    new_list_item = []
                    for ele in new_parameters:
                        new_list_item.append(list_item[parameters.index(ele)])

                    sorted_parameters = sorted(new_parameters, key=lambda x: fixed_parameters.index(x))

                    sorted_list_item = []
                    for ele in sorted_parameters:
                        sorted_list_item.append(new_list_item[new_parameters.index(ele)])

                    Title.append(title_temp)
                    Price.append(price_temp)
                    Linked.append(link)
                    Merek.append(sorted_list_item[0])
                    Varian.append(sorted_list_item[1])
                    Model.append(sorted_list_item[2])
                    Tahun.append(sorted_list_item[3])
                    Jarak_tempuh.append(sorted_list_item[4])
                    Tipe_bbm.append(sorted_list_item[5])
                    Warna.append(sorted_list_item[6])
                    Transmisi.append(sorted_list_item[7])
                    Kapasitas_mesin.append(sorted_list_item[8])

#                     print(sorted_parameters)
#                     print(sorted_list_item)

                    new_list_item.clear()
                    sorted_list_item.clear()

                else:
                    pass

                parameters.clear()
                list_item.clear()

            else:
                pass
    except:
        pass
    
    print ("check loop, page %s done" %(pages))
    pages += 1

    
data_olx = {
    "Title" : Title,
    "Link" : Linked,
    "Price" : Price,
    "Merek" : Merek,
    "Varian" : Varian,
    "Model" : Model,
    "Tahun" : Tahun,
    "Jarak Tempuh" : Jarak_tempuh,
    "Tipe BBM" : Tipe_bbm,
    "Warna" : Warna,
    "Transmisi" : Transmisi,
    "Kapasitas Mesin" : Kapasitas_mesin
}

check loop, page 1 done
check loop, page 2 done
check loop, page 3 done
check loop, page 4 done
check loop, page 5 done
check loop, page 6 done
check loop, page 7 done
check loop, page 8 done
check loop, page 9 done
check loop, page 10 done
check loop, page 11 done
check loop, page 12 done
check loop, page 13 done
check loop, page 14 done
check loop, page 15 done
check loop, page 16 done
check loop, page 17 done
check loop, page 18 done
check loop, page 19 done
check loop, page 20 done
check loop, page 21 done
check loop, page 22 done
check loop, page 23 done
check loop, page 24 done
check loop, page 25 done
check loop, page 26 done
check loop, page 27 done
check loop, page 28 done
check loop, page 29 done
check loop, page 30 done
check loop, page 31 done
check loop, page 32 done
check loop, page 33 done
check loop, page 34 done
check loop, page 35 done
check loop, page 36 done
check loop, page 37 done
check loop, page 38 done
check loop, page 39 done
check loop, page 40 done
check loo

In [3]:
df = pd.DataFrame(data_olx)
df

Title  \
0    Mercedes Benz c250 Amg + panoramic & sunroof  ...   
1    2018 Mercedes Benz C300 Coupe Facelift [9000 K...   
2               Mercedes benz/mercy tipe W211 E260 A/T   
3                       Mercedes Benz E 300 Coupe 2018   
4                           Mercedes Benz C200 AT 2012   
..                                                 ...   
961  Mercedes Benz E400 NIK 2015 Low KM Perfect Con...   
962    Mercedes benz C200 kompressor W204 ( Like C63 )   
963  MERCEDES BENZ GLE250d 2016/2017 DIESEL BLACK G...   
964             Mercedes-Benz E400 2015 putih istimewa   
965      Mercedes Benz Mercy W211 E240 Avantgarde 2004   

                                                  Link             Price  \
0    https://www.olx.co.id/item/mercedes-benz-c250-...    Rp 358.000.000   
1    https://www.olx.co.id/item/2018-mercedes-benz-...    Rp 995.000.000   
2    https://www.olx.co.id/item/mercedes-benzmercy-...    Rp 118.000.000   
3    https://www.olx.co.id/item/mercedes-benz-e-300...  Rp 1.475.000.000   
4    https://www.olx.co.id/item/mercedes-benz-c200-...    Rp 250.000.000   
..                                                 ...               ...   
961  https://www.olx.co.id/item/mercedes-benz-e400-...    Rp 615.000.000   
962  https://www.olx.co.id/item/mercedes-benz-c200-...    Rp 200.000.000   
963  https://www.olx.co.id/item/mercedes-benz-gle25...    Rp 735.000.000   
964  https://www.olx.co.id/item/mercedes-benz-e400-...    Rp 625.000.000   
965  https://www.olx.co.id/item/mercedes-benz-mercy...    Rp 215.000.000   

             Merek     Varian      Model Tahun        Jarak Tempuh Tipe BBM  \
0    Mercedes-Benz       C250    C-Class  2013    50.000-55.000 km   Bensin   
1    Mercedes-Benz       C300    C-Class  2018     5.000-10.000 km   Bensin   
2    Mercedes-Benz       E260    E-Class  2003  105.000-110.000 km   Bensin   
3    Mercedes-Benz       E300    E-Class  2018    15.000-20.000 km   Bensin   
4    Mercedes-Benz       C200    C-Class  2012    55.000-60.000 km   Bensin   
..             ...        ...        ...   ...                 ...      ...   
961  Mercedes-Benz       E400    E-Class  2015    20.000-25.000 km   Bensin   
962  Mercedes-Benz      C200K    C-Class  2009  125.000-130.000 km   Bensin   
963  Mercedes-Benz    GLE250D  GLE-Class  2016    25.000-30.000 km   Diesel   
964  Mercedes-Benz       E400    E-Class  2015    15.000-20.000 km   Bensin   
965  Mercedes-Benz  Lain-lain  Lain-lain  2004   95.000-100.000 km   Bensin   

       Warna            Transmisi    Kapasitas Mesin  
0      Putih  Automatic Triptonic  >1.500 - 2.000 cc  
1    Abu-abu            Automatic  >1.500 - 2.000 cc  
2     Silver  Automatic Triptonic  >2.000 - 3.000 cc  
3      Hitam            Automatic  >2.000 - 3.000 cc  
4      Putih            Automatic  >1.500 - 2.000 cc  
..       ...                  ...                ...  
961    Hitam  Automatic Triptonic  >2.000 - 3.000 cc  
962  Abu-abu  Automatic Triptonic  >1.500 - 2.000 cc  
963    Hitam            Automatic  >2.000 - 3.000 cc  
964    Putih  Automatic Triptonic  >2.000 - 3.000 cc  
965    Hitam            Automatic  >2.000 - 3.000 cc  

[966 rows x 12 columns]

In [7]:
len(df['Link'].unique())

655

> **After making it to dataframe, then determine the unique columns which is "Link", because each link defines for each item. Next we will drop the rows that duplicated.**

In [8]:
df1 = df.drop_duplicates(subset='Link')
len(df1['Link'].unique())

655

> **Now in our dataframe there is no duplicated item/row anymore. Each row defines different item. Then export it as csv.**

In [10]:
# export as csv
df1.to_csv('olx_mercedes_1.csv')

> **Because olx has more than 1900 items for Mercedes car and we only get 655, so I need to try scraping once again. 1900 used car / 20 items per page means that we only need ~100 loops to get all data. The reason why we only get ~600 data maybe caused by the same data that loaded in other pages, it's proved by the data that duplicated in scraping result.**

> **This may be caused too by the data that doesn't have price tag in their listing page (1st if-conditional at class "_89yzn") or the data that doesn't have paramaters which is matched with fixed_parameters (2nd if-conditional) or get error when opening the page.**


## 2nd Scraping

Trying to scrape data from pages 100-250

In [32]:
Title = []
Price = []
Linked = []
Merek = []
Varian = []
Model = []
Tahun = []
Jarak_tempuh = []
Tipe_bbm = []
Warna =[]
Transmisi = []
Kapasitas_mesin = []
fixed_parameters = ['Merek','Varian', 'Model', 'Tahun', 'Jarak tempuh', 'Tipe bahan bakar', 'Warna', 'Transmisi', 'Kapasitas mesin']

pages = 100
while pages < 250:
    try :
        html = urlopen("https://www.olx.co.id/mobil-bekas_c198/q-mercedes-benz?page=%s" %(pages)).read()
        soup = BeautifulSoup(html, "lxml")
        produk = soup.find_all("li","EIR5N")
        for p in produk:
            if (p.find("span","_89yzn") != None) :
                title_temp = p.find("span","_2tW1I").get_text()
                price_temp = p.find("span","_89yzn").get_text()
                link = "https://www.olx.co.id"+p.find("a",href=True)['href']

                html_item = urlopen(link).read()
                soup_item = BeautifulSoup(html_item, "lxml")
                parameters_item = soup_item.find_all("span","_25oXN")
                produk_item = soup_item.find_all("span","_2vNpt")

                parameters = []  
                for param in parameters_item:
                    parameters.append(param.get_text())

                list_item = []
                for item in produk_item:
                    list_item.append(item.get_text())


                new_parameters = [item for item in parameters  if item in fixed_parameters ]

                if len(new_parameters) == len(fixed_parameters):
                    new_list_item = []
                    for ele in new_parameters:
                        new_list_item.append(list_item[parameters.index(ele)])

                    sorted_parameters = sorted(new_parameters, key=lambda x: fixed_parameters.index(x))

                    sorted_list_item = []
                    for ele in sorted_parameters:
                        sorted_list_item.append(new_list_item[new_parameters.index(ele)])

                    Title.append(title_temp)
                    Price.append(price_temp)
                    Linked.append(link)
                    Merek.append(sorted_list_item[0])
                    Varian.append(sorted_list_item[1])
                    Model.append(sorted_list_item[2])
                    Tahun.append(sorted_list_item[3])
                    Jarak_tempuh.append(sorted_list_item[4])
                    Tipe_bbm.append(sorted_list_item[5])
                    Warna.append(sorted_list_item[6])
                    Transmisi.append(sorted_list_item[7])
                    Kapasitas_mesin.append(sorted_list_item[8])

#                     print(sorted_parameters)
#                     print(sorted_list_item)

                    new_list_item.clear()
                    sorted_list_item.clear()

                else:
                    pass

                parameters.clear()
                list_item.clear()

            else:
                pass
    except:
        pass
    
    print ("check loop, page %s done" %(pages))
    pages += 1

    
data_olx = {
    "Title" : Title,
    "Link" : Linked,
    "Price" : Price,
    "Merek" : Merek,
    "Varian" : Varian,
    "Model" : Model,
    "Tahun" : Tahun,
    "Jarak Tempuh" : Jarak_tempuh,
    "Tipe BBM" : Tipe_bbm,
    "Warna" : Warna,
    "Transmisi" : Transmisi,
    "Kapasitas Mesin" : Kapasitas_mesin
}

check loop, page 100 done
check loop, page 101 done
check loop, page 102 done
check loop, page 103 done
check loop, page 104 done
check loop, page 105 done
check loop, page 106 done
check loop, page 107 done
check loop, page 108 done
check loop, page 109 done
check loop, page 110 done
check loop, page 111 done
check loop, page 112 done
check loop, page 113 done
check loop, page 114 done
check loop, page 115 done
check loop, page 116 done
check loop, page 117 done
check loop, page 118 done
check loop, page 119 done
check loop, page 120 done
check loop, page 121 done
check loop, page 122 done
check loop, page 123 done
check loop, page 124 done
check loop, page 125 done
check loop, page 126 done
check loop, page 127 done
check loop, page 128 done
check loop, page 129 done
check loop, page 130 done
check loop, page 131 done
check loop, page 132 done
check loop, page 133 done
check loop, page 134 done
check loop, page 135 done
check loop, page 136 done
check loop, page 137 done
check loop, 

In [33]:
df_2 = pd.DataFrame(data_olx)
df_2

Title  \
0     Mercedes benz S 400L full option Nik 2015 pmk ...   
1     Mercedes Benz  A250 AMG 2014 2.0Turbo! Service...   
2     Mercedes Benz C200 AMG Line 2018 Hitam ISTIMEW...   
3                Mercedes Benz C200k 2008 Tinggal Pakai   
4     [CASH] MERCEDES BENZ C 200 2009 KM37RB FULLORI...   
...                                                 ...   
1213  [CASH] MERCEDES BENZ C 200 2009 KM37RB FULLORI...   
1214    Mercedes benz C200 kompressor W204 ( Like C63 )   
1215  Terawat!! Mercedes Benz C200 Kompressor 2010 K...   
1216  MERCEDES BENZ GLE250d 2016/2017 DIESEL BLACK G...   
1217          Mercedes Benz W210 E230 MT AMG AVANTGARDE   

                                                   Link             Price  \
0     https://www.olx.co.id/item/mercedes-benz-s-400...  Rp 1.295.000.000   
1     https://www.olx.co.id/item/mercedes-benz-a250-...    Rp 379.000.000   
2     https://www.olx.co.id/item/mercedes-benz-c200-...    Rp 645.000.000   
3     https://www.olx.co.id/item/mercedes-benz-c200k...    Rp 160.000.000   
4     https://www.olx.co.id/item/cash-mercedes-benz-...    Rp 185.000.000   
...                                                 ...               ...   
1213  https://www.olx.co.id/item/cash-mercedes-benz-...    Rp 185.000.000   
1214  https://www.olx.co.id/item/mercedes-benz-c200-...    Rp 200.000.000   
1215  https://www.olx.co.id/item/terawat-mercedes-be...    Rp 185.000.000   
1216  https://www.olx.co.id/item/mercedes-benz-gle25...    Rp 735.000.000   
1217  https://www.olx.co.id/item/mercedes-benz-w210-...     Rp 95.000.000   

              Merek   Varian      Model Tahun        Jarak Tempuh Tipe BBM  \
0     Mercedes-Benz   S400 L    S-Class  2016     5.000-10.000 km   Bensin   
1     Mercedes-Benz     A250    A-Class  2014    30.000-35.000 km   Bensin   
2     Mercedes-Benz     C200    C-Class  2018    15.000-20.000 km   Bensin   
3     Mercedes-Benz    C200K    C-Class  2008  120.000-125.000 km   Bensin   
4     Mercedes-Benz     C200    C-Class  2009    35.000-40.000 km   Bensin   
...             ...      ...        ...   ...                 ...      ...   
1213  Mercedes-Benz     C200    C-Class  2009    35.000-40.000 km   Bensin   
1214  Mercedes-Benz    C200K    C-Class  2009  125.000-130.000 km   Bensin   
1215  Mercedes-Benz    C200K    C-Class  2010    70.000-75.000 km   Bensin   
1216  Mercedes-Benz  GLE250D  GLE-Class  2016    25.000-30.000 km   Diesel   
1217  Mercedes-Benz     E230    E-Class  1997  120.000-125.000 km   Bensin   

        Warna            Transmisi    Kapasitas Mesin  
0       Hitam  Automatic Triptonic  >2.000 - 3.000 cc  
1       Hitam            Automatic  >1.500 - 2.000 cc  
2       Hitam  Automatic Triptonic  >1.500 - 2.000 cc  
3       Hitam  Automatic Triptonic  >1.500 - 2.000 cc  
4       Hitam  Automatic Triptonic  >1.500 - 2.000 cc  
...       ...                  ...                ...  
1213    Hitam  Automatic Triptonic  >1.500 - 2.000 cc  
1214  Abu-abu  Automatic Triptonic  >1.500 - 2.000 cc  
1215    Hitam  Automatic Triptonic  >1.500 - 2.000 cc  
1216    Hitam            Automatic  >2.000 - 3.000 cc  
1217     Biru               Manual  >2.000 - 3.000 cc  

[1218 rows x 12 columns]

In [34]:
len(df_2['Link'].unique())

110

In [35]:
df2 = df_2.drop_duplicates(subset='Link')
len(df2['Link'].unique())

110

> **we only get 110 unique items**

In [37]:
a = list(df1['Link'])
b = list(df2['Link'])

temp = []
for item in b:
    if item not in a:
        temp.append(item)
        
df2_sel = df2[df2['Link'].isin(temp)]

> **df2_sel has 15 rows, which means that from 2nd scraping we only got 15 data that not listed in first scraping process.**

In [41]:
# first concat

frames = [df1,df2_sel]
df_concat = pd.concat(frames)
df_concat

Title  \
0    Mercedes Benz c250 Amg + panoramic & sunroof  ...   
1    2018 Mercedes Benz C300 Coupe Facelift [9000 K...   
2               Mercedes benz/mercy tipe W211 E260 A/T   
3                       Mercedes Benz E 300 Coupe 2018   
4                           Mercedes Benz C200 AT 2012   
..                                                 ...   
568  MERCEDES BENZ C 200 2009 KM37RB SUPERANTIK SEP...   
683  Terawat!! Mercedes Benz C200 Kompressor 2010 K...   
685          Mercedes Benz W210 E230 MT AMG AVANTGARDE   
686  [CASH] MERCEDES BENZ C 200 Automatic Tiptronic...   
801         Mercedes Benz S350 2008 NIK 2007 ATPM W221   

                                                  Link             Price  \
0    https://www.olx.co.id/item/mercedes-benz-c250-...    Rp 358.000.000   
1    https://www.olx.co.id/item/2018-mercedes-benz-...    Rp 995.000.000   
2    https://www.olx.co.id/item/mercedes-benzmercy-...    Rp 118.000.000   
3    https://www.olx.co.id/item/mercedes-benz-e-300...  Rp 1.475.000.000   
4    https://www.olx.co.id/item/mercedes-benz-c200-...    Rp 250.000.000   
..                                                 ...               ...   
568  https://www.olx.co.id/item/mercedes-benz-c-200...    Rp 185.000.000   
683  https://www.olx.co.id/item/terawat-mercedes-be...    Rp 185.000.000   
685  https://www.olx.co.id/item/mercedes-benz-w210-...     Rp 95.000.000   
686  https://www.olx.co.id/item/cash-mercedes-benz-...    Rp 185.000.000   
801  https://www.olx.co.id/item/mercedes-benz-s350-...    Rp 310.000.000   

             Merek  Varian    Model Tahun        Jarak Tempuh Tipe BBM  \
0    Mercedes-Benz    C250  C-Class  2013    50.000-55.000 km   Bensin   
1    Mercedes-Benz    C300  C-Class  2018     5.000-10.000 km   Bensin   
2    Mercedes-Benz    E260  E-Class  2003  105.000-110.000 km   Bensin   
3    Mercedes-Benz    E300  E-Class  2018    15.000-20.000 km   Bensin   
4    Mercedes-Benz    C200  C-Class  2012    55.000-60.000 km   Bensin   
..             ...     ...      ...   ...                 ...      ...   
568  Mercedes-Benz    C200  C-Class  2009    35.000-40.000 km   Bensin   
683  Mercedes-Benz   C200K  C-Class  2010    70.000-75.000 km   Bensin   
685  Mercedes-Benz    E230  E-Class  1997  120.000-125.000 km   Bensin   
686  Mercedes-Benz    C200  C-Class  2010    65.000-70.000 km   Bensin   
801  Mercedes-Benz  S350 L  S-Class  2007    55.000-60.000 km   Bensin   

       Warna            Transmisi    Kapasitas Mesin  
0      Putih  Automatic Triptonic  >1.500 - 2.000 cc  
1    Abu-abu            Automatic  >1.500 - 2.000 cc  
2     Silver  Automatic Triptonic  >2.000 - 3.000 cc  
3      Hitam            Automatic  >2.000 - 3.000 cc  
4      Putih            Automatic  >1.500 - 2.000 cc  
..       ...                  ...                ...  
568    Hitam  Automatic Triptonic  >1.500 - 2.000 cc  
683    Hitam  Automatic Triptonic  >1.500 - 2.000 cc  
685     Biru               Manual  >2.000 - 3.000 cc  
686  Abu-abu  Automatic Triptonic  >1.500 - 2.000 cc  
801    Hitam  Automatic Triptonic          >3.000 cc  

[670 rows x 12 columns]

## 3rd Scraping

I use the data which is not scraped in this jupyter notebook (every row is unique). Need to check with the previous concated data. 

In [48]:
df = pd.read_csv('olx_mercedes_t.csv')
df = df.drop("Tipe Bodi",axis=1)
len(df["Link"].unique())

607

In [49]:
a = list(df_concat['Link'])
b = list(df['Link'])

temp = []
for item in b:
    if item not in a:
        temp.append(item)
        
df3_sel = df[df['Link'].isin(temp)]

In [50]:
len(df3_sel)

307

> It shows good result, there is 307 items that not listed in previous concated data. Next I need to concat this data with the previous one.

In [52]:
# update df_concat
frames = [df_concat,df3_sel]
df_concat = pd.concat(frames)
df_concat

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


Jarak Tempuh    Kapasitas Mesin  \
0      50.000-55.000 km  >1.500 - 2.000 cc   
1       5.000-10.000 km  >1.500 - 2.000 cc   
2    105.000-110.000 km  >2.000 - 3.000 cc   
3      15.000-20.000 km  >2.000 - 3.000 cc   
4      55.000-60.000 km  >1.500 - 2.000 cc   
..                  ...                ...   
549    30.000-35.000 km  >1.500 - 2.000 cc   
550  160.000-165.000 km  >2.000 - 3.000 cc   
566    15.000-20.000 km  >1.500 - 2.000 cc   
591    50.000-55.000 km  >1.500 - 2.000 cc   
605    15.000-20.000 km  >1.500 - 2.000 cc   

                                                  Link          Merek  \
0    https://www.olx.co.id/item/mercedes-benz-c250-...  Mercedes-Benz   
1    https://www.olx.co.id/item/2018-mercedes-benz-...  Mercedes-Benz   
2    https://www.olx.co.id/item/mercedes-benzmercy-...  Mercedes-Benz   
3    https://www.olx.co.id/item/mercedes-benz-e-300...  Mercedes-Benz   
4    https://www.olx.co.id/item/mercedes-benz-c200-...  Mercedes-Benz   
..                                                 ...            ...   
549  https://www.olx.co.id/item/mercedes-benz-e400-...  Mercedes-Benz   
550  https://www.olx.co.id/item/mercedes-benz-w124-...  Mercedes-Benz   
566  https://www.olx.co.id/item/2013-mercedes-benz-...  Mercedes-Benz   
591  https://www.olx.co.id/item/mercedes-benz-c200-...  Mercedes-Benz   
605  https://www.olx.co.id/item/mercedes-benz-slk-2...  Mercedes-Benz   

         Model             Price Tahun Tipe BBM  \
0      C-Class    Rp 358.000.000  2013   Bensin   
1      C-Class    Rp 995.000.000  2018   Bensin   
2      E-Class    Rp 118.000.000  2003   Bensin   
3      E-Class  Rp 1.475.000.000  2018   Bensin   
4      C-Class    Rp 250.000.000  2012   Bensin   
..         ...               ...   ...      ...   
549    E-Class    Rp 625.000.000  2015   Bensin   
550    E-Class     Rp 82.000.000  1990   Bensin   
566    E-Class    Rp 485.000.000  2013   Bensin   
591    C-Class    Rp 525.000.000  2015   Bensin   
605  SLK-Class    Rp 490.000.000  2012   Bensin   

                                                 Title            Transmisi  \
0    Mercedes Benz c250 Amg + panoramic & sunroof  ...  Automatic Triptonic   
1    2018 Mercedes Benz C300 Coupe Facelift [9000 K...            Automatic   
2               Mercedes benz/mercy tipe W211 E260 A/T  Automatic Triptonic   
3                       Mercedes Benz E 300 Coupe 2018            Automatic   
4                           Mercedes Benz C200 AT 2012            Automatic   
..                                                 ...                  ...   
549             Mercedes Benz E400 Matic AT Tahun 2015            Automatic   
550  Mercedes Benz W124 230e Automatic Full USDM Ar...            Automatic   
566  2013 Mercedes Benz E250 Facelift Km rendah | b...            Automatic   
591      Mercedes Benz C200 Avantgarde W205 Tahun 2015  Automatic Triptonic   
605                  Mercedes Benz SLK 25O 2012 Silver            Automatic   

     Unnamed: 0  Varian    Warna  
0           NaN    C250    Putih  
1           NaN    C300  Abu-abu  
2           NaN    E260   Silver  
3           NaN    E300    Hitam  
4           NaN    C200    Putih  
..          ...     ...      ...  
549       563.0    E400    Putih  
550       564.0    E230    Putih  
566       686.0    E250    Hitam  
591      1005.0    C200    Hitam  
605      1722.0  SLK250   Silver  

[977 rows x 13 columns]

In [57]:
df_concat = df_concat[['Title',
 'Link',
 'Price',
 'Merek',
 'Varian',
 'Model',
 'Tahun',
 'Jarak Tempuh',
 'Tipe BBM',
 'Warna',
 'Transmisi',
 'Kapasitas Mesin']]

In [59]:
# export updated df_concat
df_concat.to_csv('olx_mercedes_concat.csv')

## 4th Scraping

Trying to scrape page 1-100 again to look the differences.

In [60]:
Title = []
Price = []
Linked = []
Merek = []
Varian = []
Model = []
Tahun = []
Jarak_tempuh = []
Tipe_bbm = []
Warna =[]
Transmisi = []
Kapasitas_mesin = []
fixed_parameters = ['Merek','Varian', 'Model', 'Tahun', 'Jarak tempuh', 'Tipe bahan bakar', 'Warna', 'Transmisi', 'Kapasitas mesin']

pages = 1
while pages < 100:
    try :
        html = urlopen("https://www.olx.co.id/mobil-bekas_c198/q-mercedes-benz?page=%s" %(pages)).read()
        soup = BeautifulSoup(html, "lxml")
        produk = soup.find_all("li","EIR5N")
        for p in produk:
            if (p.find("span","_89yzn") != None) :
                title_temp = p.find("span","_2tW1I").get_text()
                price_temp = p.find("span","_89yzn").get_text()
                link = "https://www.olx.co.id"+p.find("a",href=True)['href']

                html_item = urlopen(link).read()
                soup_item = BeautifulSoup(html_item, "lxml")
                parameters_item = soup_item.find_all("span","_25oXN")
                produk_item = soup_item.find_all("span","_2vNpt")

                parameters = []  
                for param in parameters_item:
                    parameters.append(param.get_text())

                list_item = []
                for item in produk_item:
                    list_item.append(item.get_text())


                new_parameters = [item for item in parameters  if item in fixed_parameters ]

                if len(new_parameters) == len(fixed_parameters):
                    new_list_item = []
                    for ele in new_parameters:
                        new_list_item.append(list_item[parameters.index(ele)])

                    sorted_parameters = sorted(new_parameters, key=lambda x: fixed_parameters.index(x))

                    sorted_list_item = []
                    for ele in sorted_parameters:
                        sorted_list_item.append(new_list_item[new_parameters.index(ele)])

                    Title.append(title_temp)
                    Price.append(price_temp)
                    Linked.append(link)
                    Merek.append(sorted_list_item[0])
                    Varian.append(sorted_list_item[1])
                    Model.append(sorted_list_item[2])
                    Tahun.append(sorted_list_item[3])
                    Jarak_tempuh.append(sorted_list_item[4])
                    Tipe_bbm.append(sorted_list_item[5])
                    Warna.append(sorted_list_item[6])
                    Transmisi.append(sorted_list_item[7])
                    Kapasitas_mesin.append(sorted_list_item[8])

#                     print(sorted_parameters)
#                     print(sorted_list_item)

                    new_list_item.clear()
                    sorted_list_item.clear()

                else:
                    pass

                parameters.clear()
                list_item.clear()

            else:
                pass
    except:
        pass
    
    print ("check loop, page %s done" %(pages))
    pages += 1

    
data_olx = {
    "Title" : Title,
    "Link" : Linked,
    "Price" : Price,
    "Merek" : Merek,
    "Varian" : Varian,
    "Model" : Model,
    "Tahun" : Tahun,
    "Jarak Tempuh" : Jarak_tempuh,
    "Tipe BBM" : Tipe_bbm,
    "Warna" : Warna,
    "Transmisi" : Transmisi,
    "Kapasitas Mesin" : Kapasitas_mesin
}

check loop, page 1 done
check loop, page 2 done
check loop, page 3 done
check loop, page 4 done
check loop, page 5 done
check loop, page 6 done
check loop, page 7 done
check loop, page 8 done
check loop, page 9 done
check loop, page 10 done
check loop, page 11 done
check loop, page 12 done
check loop, page 13 done
check loop, page 14 done
check loop, page 15 done
check loop, page 16 done
check loop, page 17 done
check loop, page 18 done
check loop, page 19 done
check loop, page 20 done
check loop, page 21 done
check loop, page 22 done
check loop, page 23 done
check loop, page 24 done
check loop, page 25 done
check loop, page 26 done
check loop, page 27 done
check loop, page 28 done
check loop, page 29 done
check loop, page 30 done
check loop, page 31 done
check loop, page 32 done
check loop, page 33 done
check loop, page 34 done
check loop, page 35 done
check loop, page 36 done
check loop, page 37 done
check loop, page 38 done
check loop, page 39 done
check loop, page 40 done
check loo

In [61]:
df_4 = pd.DataFrame(data_olx)
df_4

Title  \
0     Mercedes Benz GLA200 AMG LINE 2015 nik 2014 SU...   
1     Mercedes Benz S280 W220 2000 Very Low Kilomete...   
2     MERCEDES BENZ A200 PROGESIF LINE 2019 WHITE  K...   
3     Mercedes benz S350 solitaire RSE tahun 2010 mu...   
4                        MERCEDES BENZ C200 AVG AT 2017   
...                                                 ...   
1068  MERCEDES BENZ C180 C200 PAJAK BARU BAYAR AC DI...   
1069  [CASH] MERCEDES BENZ C 200 Automatic Tiptronic...   
1070                  Mercedes Benz C200 Avangarde 2008   
1071  [CASH] MERCEDES BENZ C 200 2009 KM37RB FULLORI...   
1072    Mercedes benz C200 kompressor W204 ( Like C63 )   

                                                   Link           Price  \
0     https://www.olx.co.id/item/mercedes-benz-gla20...  Rp 420.000.000   
1     https://www.olx.co.id/item/mercedes-benz-s280-...  Rp 125.000.000   
2     https://www.olx.co.id/item/mercedes-benz-a200-...  Rp 640.000.000   
3     https://www.olx.co.id/item/mercedes-benz-s350-...  Rp 385.000.000   
4     https://www.olx.co.id/item/mercedes-benz-c200-...  Rp 595.000.000   
...                                                 ...             ...   
1068  https://www.olx.co.id/item/mercedes-benz-c180-...   Rp 77.000.000   
1069  https://www.olx.co.id/item/cash-mercedes-benz-...  Rp 185.000.000   
1070  https://www.olx.co.id/item/mercedes-benz-c200-...  Rp 145.000.000   
1071  https://www.olx.co.id/item/cash-mercedes-benz-...  Rp 185.000.000   
1072  https://www.olx.co.id/item/mercedes-benz-c200-...  Rp 200.000.000   

              Merek  Varian      Model Tahun        Jarak Tempuh Tipe BBM  \
0     Mercedes-Benz  GLA200  GLA-Class  2014    40.000-45.000 km   Bensin   
1     Mercedes-Benz    S280    S-Class  2000    55.000-60.000 km   Bensin   
2     Mercedes-Benz    A200    A-Class  2019          0-5.000 km   Bensin   
3     Mercedes-Benz    S350    S-Class  2010    50.000-55.000 km   Bensin   
4     Mercedes-Benz    C200    C-Class  2017    30.000-35.000 km   Bensin   
...             ...     ...        ...   ...                 ...      ...   
1068  Mercedes-Benz    C180    C-Class  2003  130.000-135.000 km   Bensin   
1069  Mercedes-Benz    C200    C-Class  2010    65.000-70.000 km   Bensin   
1070  Mercedes-Benz   C200K    C-Class  2008    80.000-85.000 km   Bensin   
1071  Mercedes-Benz    C200    C-Class  2009    35.000-40.000 km   Bensin   
1072  Mercedes-Benz   C200K    C-Class  2009  125.000-130.000 km   Bensin   

        Warna            Transmisi    Kapasitas Mesin  
0       Putih  Automatic Triptonic  >1.000 - 1.500 cc  
1       Hitam  Automatic Triptonic  >2.000 - 3.000 cc  
2       Putih  Automatic Triptonic  >1.000 - 1.500 cc  
3       Hitam  Automatic Triptonic          >3.000 cc  
4       Hitam  Automatic Triptonic  >1.500 - 2.000 cc  
...       ...                  ...                ...  
1068   Silver  Automatic Triptonic  >1.500 - 2.000 cc  
1069  Abu-abu  Automatic Triptonic  >1.500 - 2.000 cc  
1070   Coklat  Automatic Triptonic  >1.500 - 2.000 cc  
1071    Hitam  Automatic Triptonic  >1.500 - 2.000 cc  
1072  Abu-abu  Automatic Triptonic  >1.500 - 2.000 cc  

[1073 rows x 12 columns]

In [62]:
len(df_4['Link'].unique())

668

In [63]:
df4 = df_4.drop_duplicates(subset='Link')
len(df4['Link'].unique())

668

In [64]:
a = list(df_concat['Link'])
b = list(df4['Link'])

temp = []
for item in b:
    if item not in a:
        temp.append(item)
        
df4_sel = df4[df4['Link'].isin(temp)]

> **From the result we could see that the 4th scraping produces 124 items that are not listed in all previous processes, it's a good result too, need to concat them too.** 

In [65]:
# item not in df_concat
df4_sel

Title  \
5    Dijual cepat Mercedes benz Boxer 1988 tinggal ...   
9    Mercedes-Benz E300 2010.. istimewa SE INDONESI...   
16               Mercedes Benz ML 320 ML320 Matic 2002   
18                          Mercedes Benz W220 S 320 L   
23   Mercedes Benz C250 At 2013 Hitam..Super Terawa...   
..                                                 ...   
648                         Mercedes benz CLS 350 2006   
649         Mercedes Benz C240 Elegance Triptonic 2005   
757  MERCEDES BENZ C180 C200 PAJAK BARU BAYAR AC DI...   
783           Mercedes-Benz / Mercy C250 W204 2010 AMG   
922    Mercedes Benz GLC 300 Coupe AMG thn 2019 ANTIK!   

                                                  Link             Price  \
5    https://www.olx.co.id/item/dijual-cepat-merced...     Rp 30.000.000   
9    https://www.olx.co.id/item/mercedes-benz-e300-...    Rp 285.000.000   
16   https://www.olx.co.id/item/mercedes-benz-ml-32...    Rp 130.000.000   
18   https://www.olx.co.id/item/mercedes-benz-w220-...    Rp 185.000.000   
23   https://www.olx.co.id/item/mercedes-benz-c250-...    Rp 305.000.000   
..                                                 ...               ...   
648  https://www.olx.co.id/item/mercedes-benz-cls-3...    Rp 280.000.000   
649  https://www.olx.co.id/item/mercedes-benz-c240-...    Rp 135.000.000   
757  https://www.olx.co.id/item/mercedes-benz-c180-...     Rp 77.000.000   
783  https://www.olx.co.id/item/mercedes-benz-mercy...    Rp 205.000.000   
922  https://www.olx.co.id/item/mercedes-benz-glc-3...  Rp 1.125.000.000   

             Merek     Varian      Model Tahun        Jarak Tempuh Tipe BBM  \
5    Mercedes-Benz  Lain-lain  Lain-lain  1988    50.000-55.000 km   Bensin   
9    Mercedes-Benz       E300    E-Class  2010    85.000-90.000 km   Bensin   
16   Mercedes-Benz      ML320   ML-Class  2002  130.000-135.000 km   Bensin   
18   Mercedes-Benz       S320    S-Class  2000    40.000-45.000 km   Bensin   
23   Mercedes-Benz       C250    C-Class  2013    75.000-80.000 km   Bensin   
..             ...        ...        ...   ...                 ...      ...   
648  Mercedes-Benz  Lain-lain        CLS  2006    55.000-60.000 km   Bensin   
649  Mercedes-Benz       C240    C-Class  2005    90.000-95.000 km   Bensin   
757  Mercedes-Benz       C180    C-Class  2003  130.000-135.000 km   Bensin   
783  Mercedes-Benz       C250    C-Class  2010  105.000-110.000 km   Bensin   
922  Mercedes-Benz     GLC300  GLC-Class  2019          0-5.000 km   Bensin   

      Warna            Transmisi    Kapasitas Mesin  
5     Hitam               Manual  >2.000 - 3.000 cc  
9      Emas            Automatic          >3.000 cc  
16    Hijau            Automatic  >2.000 - 3.000 cc  
18     Biru  Automatic Triptonic          >3.000 cc  
23    Hitam  Automatic Triptonic  >1.500 - 2.000 cc  
..      ...                  ...                ...  
648   Hitam  Automatic Triptonic          >3.000 cc  
649  Silver  Automatic Triptonic  >2.000 - 3.000 cc  
757  Silver  Automatic Triptonic  >1.500 - 2.000 cc  
783   Hitam  Automatic Triptonic  >1.500 - 2.000 cc  
922   Putih            Automatic  >2.000 - 3.000 cc  

[124 rows x 12 columns]

In [66]:
# update df_concat

frames = [df_concat,df4_sel]
df_concat = pd.concat(frames)
df_concat

Title  \
0    Mercedes Benz c250 Amg + panoramic & sunroof  ...   
1    2018 Mercedes Benz C300 Coupe Facelift [9000 K...   
2               Mercedes benz/mercy tipe W211 E260 A/T   
3                       Mercedes Benz E 300 Coupe 2018   
4                           Mercedes Benz C200 AT 2012   
..                                                 ...   
648                         Mercedes benz CLS 350 2006   
649         Mercedes Benz C240 Elegance Triptonic 2005   
757  MERCEDES BENZ C180 C200 PAJAK BARU BAYAR AC DI...   
783           Mercedes-Benz / Mercy C250 W204 2010 AMG   
922    Mercedes Benz GLC 300 Coupe AMG thn 2019 ANTIK!   

                                                  Link             Price  \
0    https://www.olx.co.id/item/mercedes-benz-c250-...    Rp 358.000.000   
1    https://www.olx.co.id/item/2018-mercedes-benz-...    Rp 995.000.000   
2    https://www.olx.co.id/item/mercedes-benzmercy-...    Rp 118.000.000   
3    https://www.olx.co.id/item/mercedes-benz-e-300...  Rp 1.475.000.000   
4    https://www.olx.co.id/item/mercedes-benz-c200-...    Rp 250.000.000   
..                                                 ...               ...   
648  https://www.olx.co.id/item/mercedes-benz-cls-3...    Rp 280.000.000   
649  https://www.olx.co.id/item/mercedes-benz-c240-...    Rp 135.000.000   
757  https://www.olx.co.id/item/mercedes-benz-c180-...     Rp 77.000.000   
783  https://www.olx.co.id/item/mercedes-benz-mercy...    Rp 205.000.000   
922  https://www.olx.co.id/item/mercedes-benz-glc-3...  Rp 1.125.000.000   

             Merek     Varian      Model Tahun        Jarak Tempuh Tipe BBM  \
0    Mercedes-Benz       C250    C-Class  2013    50.000-55.000 km   Bensin   
1    Mercedes-Benz       C300    C-Class  2018     5.000-10.000 km   Bensin   
2    Mercedes-Benz       E260    E-Class  2003  105.000-110.000 km   Bensin   
3    Mercedes-Benz       E300    E-Class  2018    15.000-20.000 km   Bensin   
4    Mercedes-Benz       C200    C-Class  2012    55.000-60.000 km   Bensin   
..             ...        ...        ...   ...                 ...      ...   
648  Mercedes-Benz  Lain-lain        CLS  2006    55.000-60.000 km   Bensin   
649  Mercedes-Benz       C240    C-Class  2005    90.000-95.000 km   Bensin   
757  Mercedes-Benz       C180    C-Class  2003  130.000-135.000 km   Bensin   
783  Mercedes-Benz       C250    C-Class  2010  105.000-110.000 km   Bensin   
922  Mercedes-Benz     GLC300  GLC-Class  2019          0-5.000 km   Bensin   

       Warna            Transmisi    Kapasitas Mesin  
0      Putih  Automatic Triptonic  >1.500 - 2.000 cc  
1    Abu-abu            Automatic  >1.500 - 2.000 cc  
2     Silver  Automatic Triptonic  >2.000 - 3.000 cc  
3      Hitam            Automatic  >2.000 - 3.000 cc  
4      Putih            Automatic  >1.500 - 2.000 cc  
..       ...                  ...                ...  
648    Hitam  Automatic Triptonic          >3.000 cc  
649   Silver  Automatic Triptonic  >2.000 - 3.000 cc  
757   Silver  Automatic Triptonic  >1.500 - 2.000 cc  
783    Hitam  Automatic Triptonic  >1.500 - 2.000 cc  
922    Putih            Automatic  >2.000 - 3.000 cc  

[1101 rows x 12 columns]

In [67]:
len(df_concat['Link'].unique())

1101

In [69]:
# export updated df_concat
df_concat.to_csv('olx_mercedes_concat.csv')

## 5th Scraping

At this scraping process, I use looping process to load page 1-300.

In [72]:
Title = []
Price = []
Linked = []
Merek = []
Varian = []
Model = []
Tahun = []
Jarak_tempuh = []
Tipe_bbm = []
Warna =[]
Transmisi = []
Kapasitas_mesin = []
fixed_parameters = ['Merek','Varian', 'Model', 'Tahun', 'Jarak tempuh', 'Tipe bahan bakar', 'Warna', 'Transmisi', 'Kapasitas mesin']

pages = 1
while pages < 300:
    try :
        html = urlopen("https://www.olx.co.id/mobil-bekas_c198/q-mercedes-benz?page=%s" %(pages)).read()
        soup = BeautifulSoup(html, "lxml")
        produk = soup.find_all("li","EIR5N")
        for p in produk:
            if (p.find("span","_89yzn") != None) :
                title_temp = p.find("span","_2tW1I").get_text()
                price_temp = p.find("span","_89yzn").get_text()
                link = "https://www.olx.co.id"+p.find("a",href=True)['href']

                html_item = urlopen(link).read()
                soup_item = BeautifulSoup(html_item, "lxml")
                parameters_item = soup_item.find_all("span","_25oXN")
                produk_item = soup_item.find_all("span","_2vNpt")

                parameters = []  
                for param in parameters_item:
                    parameters.append(param.get_text())

                list_item = []
                for item in produk_item:
                    list_item.append(item.get_text())


                new_parameters = [item for item in parameters  if item in fixed_parameters ]

                if len(new_parameters) == len(fixed_parameters):
                    new_list_item = []
                    for ele in new_parameters:
                        new_list_item.append(list_item[parameters.index(ele)])

                    sorted_parameters = sorted(new_parameters, key=lambda x: fixed_parameters.index(x))

                    sorted_list_item = []
                    for ele in sorted_parameters:
                        sorted_list_item.append(new_list_item[new_parameters.index(ele)])

                    Title.append(title_temp)
                    Price.append(price_temp)
                    Linked.append(link)
                    Merek.append(sorted_list_item[0])
                    Varian.append(sorted_list_item[1])
                    Model.append(sorted_list_item[2])
                    Tahun.append(sorted_list_item[3])
                    Jarak_tempuh.append(sorted_list_item[4])
                    Tipe_bbm.append(sorted_list_item[5])
                    Warna.append(sorted_list_item[6])
                    Transmisi.append(sorted_list_item[7])
                    Kapasitas_mesin.append(sorted_list_item[8])

#                     print(sorted_parameters)
#                     print(sorted_list_item)

                    new_list_item.clear()
                    sorted_list_item.clear()

                else:
                    pass

                parameters.clear()
                list_item.clear()

            else:
                pass
    except:
        pass
    
    print ("check loop, page %s done" %(pages))
    pages += 1

    
data_olx = {
    "Title" : Title,
    "Link" : Linked,
    "Price" : Price,
    "Merek" : Merek,
    "Varian" : Varian,
    "Model" : Model,
    "Tahun" : Tahun,
    "Jarak Tempuh" : Jarak_tempuh,
    "Tipe BBM" : Tipe_bbm,
    "Warna" : Warna,
    "Transmisi" : Transmisi,
    "Kapasitas Mesin" : Kapasitas_mesin
}

check loop, page 1 done
check loop, page 2 done
check loop, page 3 done
check loop, page 4 done
check loop, page 5 done
check loop, page 6 done
check loop, page 7 done
check loop, page 8 done
check loop, page 9 done
check loop, page 10 done
check loop, page 11 done
check loop, page 12 done
check loop, page 13 done
check loop, page 14 done
check loop, page 15 done
check loop, page 16 done
check loop, page 17 done
check loop, page 18 done
check loop, page 19 done
check loop, page 20 done
check loop, page 21 done
check loop, page 22 done
check loop, page 23 done
check loop, page 24 done
check loop, page 25 done
check loop, page 26 done
check loop, page 27 done
check loop, page 28 done
check loop, page 29 done
check loop, page 30 done
check loop, page 31 done
check loop, page 32 done
check loop, page 33 done
check loop, page 34 done
check loop, page 35 done
check loop, page 36 done
check loop, page 37 done
check loop, page 38 done
check loop, page 39 done
check loop, page 40 done
check loo

In [73]:
df_5 = pd.DataFrame(data_olx)
df_5

Title  \
0     Mercedes Benz GLA200 AMG LINE 2017 Limited Dai...   
1                         Mercedes Benz E 400 Cabriolet   
2                Mercedes Benz C200k 2008 Tinggal Pakai   
3         Mercedes Benz C200 Avantgarde W205 Tahun 2015   
4            Mercedes benz slk 230 kompressor cabriolet   
...                                                 ...   
3496      Mercedes Benz Mercy C300 C 300 AMG 2018 Hitam   
3497  Mercedes Benz GLA200 Sport AMG Matic AT Tahun ...   
3498  NIK 2019 Mercedes Benz GLC 300 Coupe AMG KM 10...   
3499          Mercedes Benz E-Class 2011 Good Condition   
3500          Jual mercedes benz C200 Kompresor th 2009   

                                                   Link             Price  \
0     https://www.olx.co.id/item/mercedes-benz-gla20...    Rp 505.000.000   
1     https://www.olx.co.id/item/mercedes-benz-e-400...  Rp 1.050.000.000   
2     https://www.olx.co.id/item/mercedes-benz-c200k...    Rp 160.000.000   
3     https://www.olx.co.id/item/mercedes-benz-c200-...    Rp 525.000.000   
4     https://www.olx.co.id/item/mercedes-benz-slk-2...    Rp 260.000.000   
...                                                 ...               ...   
3496  https://www.olx.co.id/item/mercedes-benz-mercy...    Rp 820.000.000   
3497  https://www.olx.co.id/item/mercedes-benz-gla20...    Rp 495.000.000   
3498  https://www.olx.co.id/item/nik-2019-mercedes-b...  Rp 1.125.000.000   
3499  https://www.olx.co.id/item/mercedes-benz-e-cla...    Rp 349.000.000   
3500  https://www.olx.co.id/item/jual-mercedes-benz-...    Rp 195.000.000   

              Merek     Varian      Model Tahun        Jarak Tempuh Tipe BBM  \
0     Mercedes-Benz     GLA200  GLA-Class  2017    15.000-20.000 km   Bensin   
1     Mercedes-Benz       E400    E-Class  2015    10.000-15.000 km   Bensin   
2     Mercedes-Benz      C200K    C-Class  2008  120.000-125.000 km   Bensin   
3     Mercedes-Benz       C200    C-Class  2015    50.000-55.000 km   Bensin   
4     Mercedes-Benz  Lain-lain        SLK  1998    35.000-40.000 km   Bensin   
...             ...        ...        ...   ...                 ...      ...   
3496  Mercedes-Benz       C300    C-Class  2018    10.000-15.000 km   Bensin   
3497  Mercedes-Benz     GLA200  GLA-Class  2017    20.000-25.000 km   Bensin   
3498  Mercedes-Benz     GLC300  GLC-Class  2019          0-5.000 km   Bensin   
3499  Mercedes-Benz       E300    E-Class  2011    70.000-75.000 km   Bensin   
3500  Mercedes-Benz       C200    C-Class  2009   95.000-100.000 km   Bensin   

        Warna            Transmisi    Kapasitas Mesin  
0       Putih  Automatic Triptonic  >1.500 - 2.000 cc  
1       Hitam            Automatic  >1.000 - 1.500 cc  
2       Hitam  Automatic Triptonic  >1.500 - 2.000 cc  
3       Hitam  Automatic Triptonic  >1.500 - 2.000 cc  
4      Silver            Automatic  >2.000 - 3.000 cc  
...       ...                  ...                ...  
3496    Hitam            Automatic  >1.500 - 2.000 cc  
3497    Putih            Automatic  >1.500 - 2.000 cc  
3498    Putih            Automatic  >1.500 - 2.000 cc  
3499  Abu-abu  Automatic Triptonic  >2.000 - 3.000 cc  
3500   Silver  Automatic Triptonic  >1.500 - 2.000 cc  

[3501 rows x 12 columns]

In [74]:
len(df_5['Link'].unique())

701

In [75]:
df5 = df_5.drop_duplicates(subset='Link')
len(df5['Link'].unique())

701

In [77]:
a = list(df_concat['Link'])
b = list(df5['Link'])

temp = []
for item in b:
    if item not in a:
        temp.append(item)
        
df5_sel = df5[df5['Link'].isin(temp)]

In [78]:
# items not in df_concat
df5_sel

Title  \
7      Di jual cepat  mobil Mercedes Benz Hitam Metalik   
12    Mercedes-Benz E 250 AMG Coupe AT Putih (KM 35....   
14      Mercedes-Benz\nC207 W212 E250 Coupe At\nTh 2011   
23          Mercedes Benz MERCY E250 Thn 2017 AVG HITAM   
24                     Jual cepat Mercedes-Benz SLK 200   
...                                                 ...   
1837  NIK 2019 Mercedes Benz GLC 300 Coupe AMG KM 10...   
2376              Di Jual Mercedes Benz C250 2010 Matic   
2703  Mercedes Benz GLA200 Sport AMG Matic AT Tahun ...   
3313      Mercedes Benz Mercy C300 C 300 AMG 2018 Hitam   
3361                 Mercedes Benz S300 L Class AT 2009   

                                                   Link             Price  \
7     https://www.olx.co.id/item/di-jual-cepat-mobil...     Rp 65.000.000   
12    https://www.olx.co.id/item/mercedes-benz-e-250...    Rp 450.000.000   
14    https://www.olx.co.id/item/mercedes-benz-c207-...    Rp 565.000.000   
23    https://www.olx.co.id/item/mercedes-benz-mercy...    Rp 795.000.000   
24    https://www.olx.co.id/item/jual-cepat-mercedes...    Rp 245.000.000   
...                                                 ...               ...   
1837  https://www.olx.co.id/item/nik-2019-mercedes-b...  Rp 1.125.000.000   
2376  https://www.olx.co.id/item/di-jual-mercedes-be...    Rp 200.000.000   
2703  https://www.olx.co.id/item/mercedes-benz-gla20...    Rp 495.000.000   
3313  https://www.olx.co.id/item/mercedes-benz-mercy...    Rp 820.000.000   
3361  https://www.olx.co.id/item/mercedes-benz-s300-...    Rp 369.000.000   

              Merek     Varian      Model Tahun        Jarak Tempuh Tipe BBM  \
7     Mercedes-Benz       C230    C-Class  2007    35.000-40.000 km   Bensin   
12    Mercedes-Benz       E250    E-Class  2011    30.000-35.000 km   Bensin   
14    Mercedes-Benz       E250    E-Class  2011    30.000-35.000 km   Bensin   
23    Mercedes-Benz       E250    E-Class  2017    15.000-20.000 km   Bensin   
24    Mercedes-Benz  Lain-lain        SLK  2001  160.000-165.000 km   Bensin   
...             ...        ...        ...   ...                 ...      ...   
1837  Mercedes-Benz     GLC300  GLC-Class  2019          0-5.000 km   Bensin   
2376  Mercedes-Benz       C250    C-Class  2010    55.000-60.000 km   Bensin   
2703  Mercedes-Benz     GLA200  GLA-Class  2017    20.000-25.000 km   Bensin   
3313  Mercedes-Benz       C300    C-Class  2018    10.000-15.000 km   Bensin   
3361  Mercedes-Benz     S300 L    S-Class  2009   95.000-100.000 km   Bensin   

      Warna            Transmisi    Kapasitas Mesin  
7     Hitam            Automatic  >2.000 - 3.000 cc  
12    Putih  Automatic Triptonic  >1.500 - 2.000 cc  
14    Putih            Automatic  >2.000 - 3.000 cc  
23    Hitam  Automatic Triptonic  >1.500 - 2.000 cc  
24    Hitam  Automatic Triptonic  >2.000 - 3.000 cc  
...     ...                  ...                ...  
1837  Putih            Automatic  >1.500 - 2.000 cc  
2376  Hitam  Automatic Triptonic  >1.500 - 2.000 cc  
2703  Putih            Automatic  >1.500 - 2.000 cc  
3313  Hitam            Automatic  >1.500 - 2.000 cc  
3361  Hitam            Automatic  >2.000 - 3.000 cc  

[83 rows x 12 columns]

> **Get 83 items that are not listed before.**

In [79]:
# update df_concat

frames = [df_concat,df5_sel]
df_concat = pd.concat(frames)
df_concat

Title  \
0     Mercedes Benz c250 Amg + panoramic & sunroof  ...   
1     2018 Mercedes Benz C300 Coupe Facelift [9000 K...   
2                Mercedes benz/mercy tipe W211 E260 A/T   
3                        Mercedes Benz E 300 Coupe 2018   
4                            Mercedes Benz C200 AT 2012   
...                                                 ...   
1837  NIK 2019 Mercedes Benz GLC 300 Coupe AMG KM 10...   
2376              Di Jual Mercedes Benz C250 2010 Matic   
2703  Mercedes Benz GLA200 Sport AMG Matic AT Tahun ...   
3313      Mercedes Benz Mercy C300 C 300 AMG 2018 Hitam   
3361                 Mercedes Benz S300 L Class AT 2009   

                                                   Link             Price  \
0     https://www.olx.co.id/item/mercedes-benz-c250-...    Rp 358.000.000   
1     https://www.olx.co.id/item/2018-mercedes-benz-...    Rp 995.000.000   
2     https://www.olx.co.id/item/mercedes-benzmercy-...    Rp 118.000.000   
3     https://www.olx.co.id/item/mercedes-benz-e-300...  Rp 1.475.000.000   
4     https://www.olx.co.id/item/mercedes-benz-c200-...    Rp 250.000.000   
...                                                 ...               ...   
1837  https://www.olx.co.id/item/nik-2019-mercedes-b...  Rp 1.125.000.000   
2376  https://www.olx.co.id/item/di-jual-mercedes-be...    Rp 200.000.000   
2703  https://www.olx.co.id/item/mercedes-benz-gla20...    Rp 495.000.000   
3313  https://www.olx.co.id/item/mercedes-benz-mercy...    Rp 820.000.000   
3361  https://www.olx.co.id/item/mercedes-benz-s300-...    Rp 369.000.000   

              Merek  Varian      Model Tahun        Jarak Tempuh Tipe BBM  \
0     Mercedes-Benz    C250    C-Class  2013    50.000-55.000 km   Bensin   
1     Mercedes-Benz    C300    C-Class  2018     5.000-10.000 km   Bensin   
2     Mercedes-Benz    E260    E-Class  2003  105.000-110.000 km   Bensin   
3     Mercedes-Benz    E300    E-Class  2018    15.000-20.000 km   Bensin   
4     Mercedes-Benz    C200    C-Class  2012    55.000-60.000 km   Bensin   
...             ...     ...        ...   ...                 ...      ...   
1837  Mercedes-Benz  GLC300  GLC-Class  2019          0-5.000 km   Bensin   
2376  Mercedes-Benz    C250    C-Class  2010    55.000-60.000 km   Bensin   
2703  Mercedes-Benz  GLA200  GLA-Class  2017    20.000-25.000 km   Bensin   
3313  Mercedes-Benz    C300    C-Class  2018    10.000-15.000 km   Bensin   
3361  Mercedes-Benz  S300 L    S-Class  2009   95.000-100.000 km   Bensin   

        Warna            Transmisi    Kapasitas Mesin  
0       Putih  Automatic Triptonic  >1.500 - 2.000 cc  
1     Abu-abu            Automatic  >1.500 - 2.000 cc  
2      Silver  Automatic Triptonic  >2.000 - 3.000 cc  
3       Hitam            Automatic  >2.000 - 3.000 cc  
4       Putih            Automatic  >1.500 - 2.000 cc  
...       ...                  ...                ...  
1837    Putih            Automatic  >1.500 - 2.000 cc  
2376    Hitam  Automatic Triptonic  >1.500 - 2.000 cc  
2703    Putih            Automatic  >1.500 - 2.000 cc  
3313    Hitam            Automatic  >1.500 - 2.000 cc  
3361    Hitam            Automatic  >2.000 - 3.000 cc  

[1184 rows x 12 columns]

In [80]:
len(df_concat['Link'].unique())

1184

In [81]:
# export updated df_concat
df_concat.to_csv('olx_mercedes_concat.csv')

## 6th Scraping

Try to edit script which is remove if-conditional price != None

In [86]:
Title = []
Price = []
Linked = []
Merek = []
Varian = []
Model = []
Tahun = []
Jarak_tempuh = []
Tipe_bbm = []
Warna =[]
Transmisi = []
Kapasitas_mesin = []
fixed_parameters = ['Merek','Varian', 'Model', 'Tahun', 'Jarak tempuh', 'Tipe bahan bakar', 'Warna', 'Transmisi', 'Kapasitas mesin']

pages = 1
while pages < 300:
    try :
        html = urlopen("https://www.olx.co.id/mobil-bekas_c198/q-mercedes-benz?page=%s" %(pages)).read()
        soup = BeautifulSoup(html, "lxml")
        produk = soup.find_all("li","EIR5N")
        for p in produk:
            title_temp = p.find("span","_2tW1I").get_text()
            # remove the condition when no price in listing page
#             price_temp = p.find("span","_89yzn").get_text()
            link = "https://www.olx.co.id"+p.find("a",href=True)['href']

            html_item = urlopen(link).read()
            soup_item = BeautifulSoup(html_item, "lxml")
            parameters_item = soup_item.find_all("span","_25oXN")
            produk_item = soup_item.find_all("span","_2vNpt")
            # move the price to here, and get it from item page
            price_temp = soup_item.find("span","_2xKfz").get_text()

            parameters = []  
            for param in parameters_item:
                parameters.append(param.get_text())

            list_item = []
            for item in produk_item:
                list_item.append(item.get_text())


            new_parameters = [item for item in parameters  if item in fixed_parameters ]

            if len(new_parameters) == len(fixed_parameters):
                new_list_item = []
                for ele in new_parameters:
                    new_list_item.append(list_item[parameters.index(ele)])

                sorted_parameters = sorted(new_parameters, key=lambda x: fixed_parameters.index(x))

                sorted_list_item = []
                for ele in sorted_parameters:
                    sorted_list_item.append(new_list_item[new_parameters.index(ele)])

                Title.append(title_temp)
                Price.append(price_temp)
                Linked.append(link)
                Merek.append(sorted_list_item[0])
                Varian.append(sorted_list_item[1])
                Model.append(sorted_list_item[2])
                Tahun.append(sorted_list_item[3])
                Jarak_tempuh.append(sorted_list_item[4])
                Tipe_bbm.append(sorted_list_item[5])
                Warna.append(sorted_list_item[6])
                Transmisi.append(sorted_list_item[7])
                Kapasitas_mesin.append(sorted_list_item[8])

#                 print(sorted_parameters)
#                 print(sorted_list_item)

                new_list_item.clear()
                sorted_list_item.clear()

            else:
                pass

            parameters.clear()
            list_item.clear()

    except:
        pass
    
    print ("check loop, page %s done" %(pages))
    pages += 1

    
data_olx = {
    "Title" : Title,
    "Link" : Linked,
    "Price" : Price,
    "Merek" : Merek,
    "Varian" : Varian,
    "Model" : Model,
    "Tahun" : Tahun,
    "Jarak Tempuh" : Jarak_tempuh,
    "Tipe BBM" : Tipe_bbm,
    "Warna" : Warna,
    "Transmisi" : Transmisi,
    "Kapasitas Mesin" : Kapasitas_mesin
}

check loop, page 1 done
check loop, page 2 done
check loop, page 3 done
check loop, page 4 done
check loop, page 5 done
check loop, page 6 done
check loop, page 7 done
check loop, page 8 done
check loop, page 9 done
check loop, page 10 done
check loop, page 11 done
check loop, page 12 done
check loop, page 13 done
check loop, page 14 done
check loop, page 15 done
check loop, page 16 done
check loop, page 17 done
check loop, page 18 done
check loop, page 19 done
check loop, page 20 done
check loop, page 21 done
check loop, page 22 done
check loop, page 23 done
check loop, page 24 done
check loop, page 25 done
check loop, page 26 done
check loop, page 27 done
check loop, page 28 done
check loop, page 29 done
check loop, page 30 done
check loop, page 31 done
check loop, page 32 done
check loop, page 33 done
check loop, page 34 done
check loop, page 35 done
check loop, page 36 done
check loop, page 37 done
check loop, page 38 done
check loop, page 39 done
check loop, page 40 done
check loo

In [87]:
df_new_script = pd.DataFrame(data_olx)
df_new_script

Title  \
0                         Mercedes Benz C250 2012 Mercy   
1     Mercedes Benz E400 AMG Dynamic Edition 2015 is...   
2     Mercedes Benz GLC 300 Coupe AMG 2017 (Jarang ada)   
3     Mercedes Benz E200 Low KM Terawat Full Ori (Pr...   
4                               Mercedes Benz E230 w124   
...                                                 ...   
4660                       Mercedes benz C 200 CGI 2014   
4661                Mercedes benz E 400 tahun 2015/2016   
4662  Mercedes benz C200 Avantgarde 2016 Perfect con...   
4663                  Mercedes Benz E280 7G Tronic 2008   
4664  2017 Mercedes Benz S400 ATPM NIK 2016 [7000 Mi...   

                                                   Link             Price  \
0     https://www.olx.co.id/item/mercedes-benz-c250-...    Rp 295.000.000   
1     https://www.olx.co.id/item/mercedes-benz-e400-...    Rp 585.000.000   
2     https://www.olx.co.id/item/mercedes-benz-glc-3...    Rp 980.000.000   
3     https://www.olx.co.id/item/mercedes-benz-e200-...    Rp 280.000.000   
4     https://www.olx.co.id/item/mercedes-benz-e230-...     Rp 35.000.000   
...                                                 ...               ...   
4660  https://www.olx.co.id/item/mercedes-benz-c-200...    Rp 330.000.000   
4661  https://www.olx.co.id/item/mercedes-benz-e-400...    Rp 690.000.000   
4662  https://www.olx.co.id/item/mercedes-benz-c200-...    Rp 549.000.000   
4663  https://www.olx.co.id/item/mercedes-benz-e280-...    Rp 215.000.000   
4664  https://www.olx.co.id/item/2017-mercedes-benz-...  Rp 1.350.000.000   

              Merek  Varian      Model Tahun        Jarak Tempuh Tipe BBM  \
0     Mercedes-Benz    C250    C-Class  2012    55.000-60.000 km   Bensin   
1     Mercedes-Benz    E400    E-Class  2015    20.000-25.000 km   Bensin   
2     Mercedes-Benz  GLC300  GLC-Class  2017    15.000-20.000 km   Bensin   
3     Mercedes-Benz    E200    E-Class  2012    10.000-15.000 km   Bensin   
4     Mercedes-Benz    E230    E-Class  1991  130.000-135.000 km   Bensin   
...             ...     ...        ...   ...                 ...      ...   
4660  Mercedes-Benz    C200    C-Class  2014    30.000-35.000 km   Bensin   
4661  Mercedes-Benz    E400    E-Class  2015    30.000-35.000 km   Bensin   
4662  Mercedes-Benz    C200    C-Class  2016    30.000-35.000 km   Bensin   
4663  Mercedes-Benz    E280    E-Class  2008    30.000-35.000 km   Bensin   
4664  Mercedes-Benz  S400 L    S-Class  2016    10.000-15.000 km   Bensin   

       Warna            Transmisi    Kapasitas Mesin  
0      Hitam  Automatic Triptonic  >1.500 - 2.000 cc  
1     Silver            Automatic  >2.000 - 3.000 cc  
2      Hitam            Automatic          >3.000 cc  
3      Hitam            Automatic  >1.500 - 2.000 cc  
4     Silver               Manual  >2.000 - 3.000 cc  
...      ...                  ...                ...  
4660   Hitam            Automatic  >1.500 - 2.000 cc  
4661   Hitam  Automatic Triptonic  >2.000 - 3.000 cc  
4662   Hitam  Automatic Triptonic  >1.500 - 2.000 cc  
4663   Hitam  Automatic Triptonic  >2.000 - 3.000 cc  
4664   Hitam  Automatic Triptonic          >3.000 cc  

[4665 rows x 12 columns]

In [88]:
len(df_new_script['Link'].unique())

734

In [89]:
df6 = df_new_script.drop_duplicates(subset='Link')
len(df6['Link'].unique())

734

In [90]:
a = list(df_concat['Link'])
b = list(df6['Link'])

temp = []
for item in b:
    if item not in a:
        temp.append(item)
        
df6_sel = df6[df6['Link'].isin(temp)]

In [91]:
df6_sel

Title  \
4                               Mercedes Benz E230 w124   
5     Mercedes Benz C300 AMG 2018 KM 8Rb Pjk 1Thn Pn...   
6     Mercedes Benz C300 AMG 2018 KM 8Rb Pjk 1Thn Pn...   
7                 Mercedes Benz E300 Avantgarde AT 2010   
8     Mercedes Benz C300 AMG 2018 KM 8Rb Pjk 1Thn Pn...   
...                                                 ...   
2822   Mercedes Benz C200 Avangard AT tahun 2009 low KM   
2905                  Mercedes Benz C200 CGI 2014 Black   
3165                         Mercedes Benz C180 Esprite   
3992  Mercedes benz W211 E260 / Mercy E240 2005 Blac...   
3993  Dijual Cepat berkualitas  W220 Mercedes Benz 2...   

                                                   Link           Price  \
4     https://www.olx.co.id/item/mercedes-benz-e230-...   Rp 35.000.000   
5     https://www.olx.co.id/item/mercedes-benz-c300-...  Rp 785.000.000   
6     https://www.olx.co.id/item/mercedes-benz-c300-...  Rp 785.000.000   
7     https://www.olx.co.id/item/mercedes-benz-e300-...  Rp 315.000.000   
8     https://www.olx.co.id/item/mercedes-benz-c300-...  Rp 785.000.000   
...                                                 ...             ...   
2822  https://www.olx.co.id/item/mercedes-benz-c200-...  Rp 179.000.000   
2905  https://www.olx.co.id/item/mercedes-benz-c200-...  Rp 295.000.000   
3165  https://www.olx.co.id/item/mercedes-benz-c180-...   Rp 88.000.000   
3992  https://www.olx.co.id/item/mercedes-benz-w211-...  Rp 137.000.000   
3993  https://www.olx.co.id/item/dijual-cepat-berkua...  Rp 175.000.000   

              Merek Varian    Model Tahun        Jarak Tempuh Tipe BBM  \
4     Mercedes-Benz   E230  E-Class  1991  130.000-135.000 km   Bensin   
5     Mercedes-Benz   C300  C-Class  2018     5.000-10.000 km   Bensin   
6     Mercedes-Benz   C300  C-Class  2018     5.000-10.000 km   Bensin   
7     Mercedes-Benz   E300  E-Class  2010    35.000-40.000 km   Bensin   
8     Mercedes-Benz   C300  C-Class  2018     5.000-10.000 km   Bensin   
...             ...    ...      ...   ...                 ...      ...   
2822  Mercedes-Benz   C200  C-Class  2009    50.000-55.000 km   Bensin   
2905  Mercedes-Benz   C200  C-Class  2014    30.000-35.000 km   Bensin   
3165  Mercedes-Benz   C180  C-Class  1994  150.000-155.000 km   Bensin   
3992  Mercedes-Benz   E260  E-Class  2005    85.000-90.000 km   Bensin   
3993  Mercedes-Benz   S280  S-Class  2003    45.000-50.000 km   Bensin   

       Warna            Transmisi    Kapasitas Mesin  
4     Silver               Manual  >2.000 - 3.000 cc  
5      Hitam            Automatic  >1.500 - 2.000 cc  
6      Hitam            Automatic  >1.500 - 2.000 cc  
7     Silver            Automatic  >2.000 - 3.000 cc  
8      Hitam            Automatic  >1.500 - 2.000 cc  
...      ...                  ...                ...  
2822   Hitam            Automatic  >1.500 - 2.000 cc  
2905   Hitam  Automatic Triptonic  >1.500 - 2.000 cc  
3165  Silver               Manual  >1.500 - 2.000 cc  
3992   Hitam  Automatic Triptonic  >2.000 - 3.000 cc  
3993   Hitam            Automatic  >2.000 - 3.000 cc  

[262 rows x 12 columns]

> **It's very good result, it gives 262 items that are not listed before in the previous scraping processes.**

In [93]:
# update df_concat

frames = [df_concat,df6_sel]
df_concat = pd.concat(frames)
df_concat

Title  \
0     Mercedes Benz c250 Amg + panoramic & sunroof  ...   
1     2018 Mercedes Benz C300 Coupe Facelift [9000 K...   
2                Mercedes benz/mercy tipe W211 E260 A/T   
3                        Mercedes Benz E 300 Coupe 2018   
4                            Mercedes Benz C200 AT 2012   
...                                                 ...   
2822   Mercedes Benz C200 Avangard AT tahun 2009 low KM   
2905                  Mercedes Benz C200 CGI 2014 Black   
3165                         Mercedes Benz C180 Esprite   
3992  Mercedes benz W211 E260 / Mercy E240 2005 Blac...   
3993  Dijual Cepat berkualitas  W220 Mercedes Benz 2...   

                                                   Link             Price  \
0     https://www.olx.co.id/item/mercedes-benz-c250-...    Rp 358.000.000   
1     https://www.olx.co.id/item/2018-mercedes-benz-...    Rp 995.000.000   
2     https://www.olx.co.id/item/mercedes-benzmercy-...    Rp 118.000.000   
3     https://www.olx.co.id/item/mercedes-benz-e-300...  Rp 1.475.000.000   
4     https://www.olx.co.id/item/mercedes-benz-c200-...    Rp 250.000.000   
...                                                 ...               ...   
2822  https://www.olx.co.id/item/mercedes-benz-c200-...    Rp 179.000.000   
2905  https://www.olx.co.id/item/mercedes-benz-c200-...    Rp 295.000.000   
3165  https://www.olx.co.id/item/mercedes-benz-c180-...     Rp 88.000.000   
3992  https://www.olx.co.id/item/mercedes-benz-w211-...    Rp 137.000.000   
3993  https://www.olx.co.id/item/dijual-cepat-berkua...    Rp 175.000.000   

              Merek Varian    Model Tahun        Jarak Tempuh Tipe BBM  \
0     Mercedes-Benz   C250  C-Class  2013    50.000-55.000 km   Bensin   
1     Mercedes-Benz   C300  C-Class  2018     5.000-10.000 km   Bensin   
2     Mercedes-Benz   E260  E-Class  2003  105.000-110.000 km   Bensin   
3     Mercedes-Benz   E300  E-Class  2018    15.000-20.000 km   Bensin   
4     Mercedes-Benz   C200  C-Class  2012    55.000-60.000 km   Bensin   
...             ...    ...      ...   ...                 ...      ...   
2822  Mercedes-Benz   C200  C-Class  2009    50.000-55.000 km   Bensin   
2905  Mercedes-Benz   C200  C-Class  2014    30.000-35.000 km   Bensin   
3165  Mercedes-Benz   C180  C-Class  1994  150.000-155.000 km   Bensin   
3992  Mercedes-Benz   E260  E-Class  2005    85.000-90.000 km   Bensin   
3993  Mercedes-Benz   S280  S-Class  2003    45.000-50.000 km   Bensin   

        Warna            Transmisi    Kapasitas Mesin  
0       Putih  Automatic Triptonic  >1.500 - 2.000 cc  
1     Abu-abu            Automatic  >1.500 - 2.000 cc  
2      Silver  Automatic Triptonic  >2.000 - 3.000 cc  
3       Hitam            Automatic  >2.000 - 3.000 cc  
4       Putih            Automatic  >1.500 - 2.000 cc  
...       ...                  ...                ...  
2822    Hitam            Automatic  >1.500 - 2.000 cc  
2905    Hitam  Automatic Triptonic  >1.500 - 2.000 cc  
3165   Silver               Manual  >1.500 - 2.000 cc  
3992    Hitam  Automatic Triptonic  >2.000 - 3.000 cc  
3993    Hitam            Automatic  >2.000 - 3.000 cc  

[1446 rows x 12 columns]

In [94]:
len(df_concat['Link'].unique())

1446

> **Now I have 1446 mercedes data from 1978 that listed in olx.co.id (~74% data). It may extend more, but I think it's enough for now. May be could be improved with more mercedes data or scrape other brand data.**

In [95]:
# export updated df_concat
df_concat.to_csv('olx_mercedes_concat.csv')